### ISSUE TITLE
Open in Own Window - Inconsistent 

### ISSUE THREAD

**Issue by siskyn** [threethan/LightningLauncher#21]
Setting 2d android apps to open in their own window is... Weird. 

1) needs to be set twice to remain set. 
2) checking the setting will turn it off. 

Launcher version 5.1
See the attached video. 
[https://youtu.be/zNjOldqcqAI](https://youtu.be/zNjOldqcqAI)

**Comment by threethan** [comment]
Pushed an update that _should_ have just fixed that

**Comment by siskyn** [comment]
I'm sorry to say the issue seems to remain unchanged on 5.1.2

**Comment by threethan** [comment]
Give it another try in 5.2, will reopen again if it's not actually fixed

**Comment by siskyn** [comment]
Updated to 5.2.0 and I can confirm the following:

1. needs to be set twice to remain set. - Fixed
2. checking the setting will turn it off. - Not fixed

Steps to reproduce:
1. Set an app to open in own window.
2. Launch the app
3. Close the app
4. In the launcher, go back to the app's settings. We can see "open in own window" is still set to on.
5. Click "confirm". 
6. Open the app's settings again. "open in own window" will be off.

If we don't do steps 2 and 3, the bug won't happen.
If at step 5 we click outside the pop-up to close it, instead of clicking confirm, the bug will not happen.


A second way to reproduce is:
1. Set an app to open in own window.
2. Close the launcher from the X in the top right
3. Open the launcher
4. Open the app's settings again. "open in own window" will be off.

So to sumarize:
After setting an app to open in it's own window, restarting the launcher will imediately set it to not run in own window. If instead of restarting we open the app and close it then go back to the launcher, It's still set to open in own window but clicking "confirm" in that pop-up menu will cause the bug.

**Comment by threethan** [comment]
That helped me catch it, and the issue is properly fixed in 5.3

**Comment by siskyn** [comment]
I'm sorry to say it appears to still be unresolved after the update. Please let me know if there's anything i can do to help.

**Comment by siskyn** [comment]
I forked the project to play around and "fixed" of sorts, both issues.

1. needs to be set twice to remain set. -> Go to LauncherService > move the " launchModeSwitch.setChecked" line to after the listener is defined.
2. The open in own window toggle becoming unset after clicking confirm again -> comment out the "launcherActivity.refreshInterfaceAll()"  line for the confirm dialog button. I'm not entirely sure what this does, my experience goes more towards WinForms C# coding than android, but it seems that it was sometimes stealthily resetting the toggle to off even though it appeared on ? And then was getting saved as off.

Here's the full LauncherService file with my edits for reference https://pastebin.com/Tw2HZ4rK

I'm not quite comfortable calling it a fix to just comment out a function that I don't understand even though I don't see any negative effects to it, but hopefully this gives you a good starting point.

**Comment by siskyn** [comment]
@threethan Sorry for the tag, just to make sure this isn't missed as the issue is closed.

**Comment by threethan** [comment]
Prior to the last update, I was able to replicate this by following your instructions. As of the last update, I cannot.
If you could provide instructions which work on 5.3 I can reopen and hopefully fix the issue you're having.

Here are the issues I've fixed so far:
- The warning dialog could desync the switch from the actual stored state (5.1.2)
- Some possibly related threading issues (5.2)
- The list of "launchOutApps" would not be correctly filled upon an interface refresh, unless another unrelated function was called first (5.3)

I believe your second method to reproduce was fixed in 5.3. I didn't fully understand your first method but I believe that was fixed in 5.3.1

**Comment by siskyn** [comment]
I made a video showing the issues, Please see below for timestamps. Link: [https://youtu.be/pxOswgFJ4eE](https://youtu.be/pxOswgFJ4eE)

00:00 - 00:20 -> Showing the launcher version
00:20 - 00:36 -> Replicating the number 1 issue "needs to be set twice to remain set."
00:36 -01:00 -> Replicating the following steps for Fx Explorer:
1. Set an app to open in own window.
2. Launch the app
3. Close the app
4. In the launcher, go back to the app's settings. We can see "open in own window" is still set to on.
5. Click "confirm".
6. Open the app's settings again. "open in own window" will be off.

01:00 - 01:20-> Replicating the following steps for FTP Server:
1. Set an app to open in own window.
2. Close the launcher from the X in the top right
3. Open the launcher
4. Open the app's settings again. "open in own window" will correctly be on. Click confirm
5. Open the app's settings again. "open in own window" will  be off.

01:20 -  2:50 -> Attempting to replicate the situations in my modified version of the launcher but it crashes a few times.
2:50 - 3:40 -> Replicating the steps from time index 00:36 -01:00 and the bug doesn't occur
3:40 - 3:55 -> Replicating the steps from time index 00:20 - 00:36 and the bug doesn't occur



Notes:
- While In the video, I opened the test version with my edits from the unknown sources list, and the official version from the explore shortcut, I tested off video and the behaviour is identical in the official version even if I open it from unknown sources. This doesn't seem to be a factor.

**Comment by threethan** [comment]
The fact that's I'm unable to replicate makes this seem like a race condition. Could you try the build in this zip?
[App-debug.zip](https://github.com/threethan/LightningLauncher/files/12772574/App-debug.zip)

**Comment by threethan** [comment]
Or if that doesn't work either: 
[App-debug2.zip](https://github.com/threethan/LightningLauncher/files/12772613/App-debug2.zip)

Thanks for the detailed steps

**Comment by siskyn** [comment]
One thing I want to mention is that after setting it twice, it does work and as long as i don't open the menu to see the state of that toggle, it correctly remembers the value. it only becomes disabled when I open the menu and see it visually enabled and click confirm. At that time it actually saves it as disabled.

I noticed I need to be very careful how I replicate... its very sensitive to having a"clean slate". After every test I need to reset to initial conditions by turning off run in own window and restarting the launcher.

app-debug:
1.  the first time after opening the launcher it still needs to be set twice. further edits (while launcher still open) are remembered in the first attempt. if after first opening the launcher I set it to run in own window only once and then run the app, it runs in the same window so the value is not correctly stored.
2. for steps 00:36 -01:00 in the last video, the issue persists.
3. for steps 01:00 - 01:20, the issue persists.

app-debug2:
sames as the first app-debug for all 3 points.

perhaps it would help to add logging to file functionality with a toggle for it in the settings menu, to see what's happening?

Sorry this proves such a tricky and persistant issue, and thank you for taking so much time looking into it, I really appreciate it.

**Comment by threethan** [comment]
Very much appreciate the specific feedback. I'll look into this more properly when I have a bit more free time.

**Comment by threethan** [comment]
Please check if this is resolved in 6.0.0

**Comment by Rami-Pastrami** [comment]
I have been also experiencing the issue on version 5.3.1 from sidequest. I uninstalled it, installed 6.0 (also reinstalled the shortcut for messenger). Issue still persists on this end. I am using a Quest pro, OS v57

**Comment by siskyn** [comment]
Interesting. After the automatic update, it now appears fully fixed for me.

**Comment by threethan** [comment]
The original issue this thread seems to have been resolved as well as a few related ones, so I'm going to close it for the sake of specificity.

@Rami-Pastrami , feel free to make a separate issue with the specific problem you're having. If possible, I'd appreciate specific steps for how to replicate it similar to those @siskyn provided.
